streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java

```diff
@@ -213,6 +224,177 @@ public void testUpdate() throws IOException {
+                final Bytes wrap = Bytes.wrap(integerSerializer.serialize(null, data.key()));
+                return WindowKeySchema.toStoreKeyBinary(new Windowed<>(wrap, data.window()), 1).get();
+            }
+
+            @Override
+            public void close() {
+
+            }
+        };
+
+        final MockRestoreConsumer<Windowed<Integer>, Long> restoreStateConsumer = new MockRestoreConsumer<>(windowedChangelogKeySerializer, new LongSerializer());
+        final StoreChangelogReader changelogReader = new StoreChangelogReader(
+            restoreStateConsumer,
+            Duration.ZERO,
+            stateRestoreListener,
+            new LogContext("standby-task-test ")
+        );
+
+        restoreStateConsumer.updatePartitions(changelogName, Utils.mkList(
+            new PartitionInfo(changelogName, 0, Node.noNode(), new Node[0], new Node[0])));
+        final InternalStreamsBuilder builder = new InternalStreamsBuilder(new InternalTopologyBuilder());
+        final String storeName = "windowed-store";
+        builder.stream(Collections.singleton("topic"), new ConsumedInternal<>()).groupByKey().windowedBy(TimeWindows.of(60_000).until(120_000)).count(Materialized.as(storeName));
+
+        final StreamsConfig config = createConfig(baseDir);
+        final InternalTopologyBuilder internalTopologyBuilder = InternalStreamsBuilderTest.internalTopologyBuilder(builder);
+        final ProcessorTopology topology = internalTopologyBuilder.setApplicationId(applicationId).build(0);
+
+        final StandbyTask task = new StandbyTask(taskId, partitions, topology, consumer, changelogReader, config,
+            new MockStreamsMetrics(new Metrics()), stateDirectory);
+
+        task.initializeStateStores();
+
+        restoreStateConsumer.assign(partitions);
+
+        for (ConsumerRecord<Windowed<Integer>, Long> record : Arrays.asList(
+            new ConsumerRecord<>(changelogName, 1, 10, 0L, TimestampType.CREATE_TIME, 0L, 0, 0, new Windowed<>(1, new TimeWindow(0, 60_000)), 100L),
+            new ConsumerRecord<>(changelogName, 1, 20, 60_000L, TimestampType.CREATE_TIME, 0L, 0, 0, new Windowed<>(2, new TimeWindow(60_000, 120_000)), 100L),
+            new ConsumerRecord<>(changelogName, 1, 30, 120_000L, TimestampType.CREATE_TIME, 0L, 0, 0, new Windowed<>(3, new TimeWindow(120_000, 180_000)), 100L),
+            new ConsumerRecord<>(changelogName, 1, 40, 180_000L, TimestampType.CREATE_TIME, 0L, 0, 0, new Windowed<>(4, new TimeWindow(180_000, 240_000)), 100L)
+        )) {
+            restoreStateConsumer.bufferRecord(record);
+        }
+
+        restoreStateConsumer.seekToBeginning(partitions);
+
+        final List<ConsumerRecord<byte[], byte[]>> records = restoreStateConsumer.poll(Duration.ofMillis(100)).records(topicPartition);
+
+        final List<ConsumerRecord<byte[], byte[]>> remaining1 = task.update(topicPartition, records);
+
+        // the first record's window should have expired.
```

Comment:
cannot follow: committed offset is zero, thus, just no restore happens. How is this related to window expiration?